### PR TITLE
Off chain warp messages

### PR DESCRIPTION
**Pull request by cam-schultz**
## Why this should be merged
Adds support for off-chain Warp messages that upgrade the `TeleporterMessenger` version registered in `TeleporterRegistry` on a given chain. Does **not** provide support for generic off-chain messages that invoke arbitrary methods of arbitrary contracts. This design can be extended to support that in the future.

Fixes https://github.com/ava-labs/awm-relayer/issues/12

## How this works
Functional changes:
- Adds package `offchainregistry` that implements `MessageManager`. This type specifically implements off-chain Warp messages that call the `addProtocolVersion` method of `TeleporterRegistry`
- Adds a configuration for this message protocol consisting of the `TeleporterRegistry` contract address. Each chain that is configured to use this message protocol must provide their `TeleporterRegistry` address.

Additional changes:
- Updates `subnet-evm` to `v0.5.11`
- Updates `avalanchego` to `v1.10.18`
  - This version of `avalanchego` broke the relayer's integration with the App Request test network. Rather than debugging something we're likely to deprecate anyway (see https://github.com/ava-labs/awm-relayer/issues/119), I elected to switch to calling `createSignedMessage` instead of `createSignedMessageAppRequest` in this PR. We should follow up and remove all of the App Request code from the relayer entirely.
- Simplifies the MessageManager` interface
- Removes unneeded `WarpMessageInfo` type

## How this was tested
- Added `teleporter_registry.go` E2E test
- CI

## How is this documented
N/A

### CODE REVIEW

File: go.mod
Comment (geoff-vball) [inline go.mod:9]: I have so many questions about this tag... How did we get `v0.1.1-0`? I don't see that tag in the teleporter repo. Also why is there a dash? Was this supposed to be an rc?

File: messages/off-chain-registry/message_manager.go
Comment (geoff-vball) [inline messages/off-chain-registry/message_manager.go:18]: nit: could be `common.Address{}`

File: messages/off-chain-registry/message_manager.go
Comment (geoff-vball) [inline messages/off-chain-registry/message_manager.go:33]: This is unused. I don't think we need to keep the function signature for `NewMessageManager` the same, so I'd just remove it. If we do need to keep it for some reason, just rename it to `_`

File: messages/off-chain-registry/message_manager.go
Comment (geoff-vball) [inline messages/off-chain-registry/message_manager.go:75]: Can you add a comment saying we only include one warp message at a time, so the warp message index is always 0?

File: messages/off-chain-registry/message_manager.go
Comment (geoff-vball) [inline messages/off-chain-registry/message_manager.go:85]: I think we should probably test the transaction before we actually send it. Right now if the relayer restarts, it'll try to send the message again, which will revert, right?

File: go.mod
Comment (cam-schultz) [reply go.mod:9]: This includes the changes in https://github.com/ava-labs/teleporter/pull/256. `d92ef8e89b7a` is the first commit of that PR - I'll go back and fetch the latest commit once that PR is merged.

From what I've observed, when fetching a non-tagged commit using `go get`, go.mod will take the latest semver release and automatically increment the bugfix version before appending the timestamp and commit hash.

The changes to go.mod here were made by running `go get github.com/ava-labs/teleporter@d92ef8e89b7a`

File: messages/off-chain-registry/message_manager.go
Comment (cam-schultz) [reply messages/off-chain-registry/message_manager.go:18]: Yes, but I think it's a bit clearer that this is the zero address and not an uninitalized value. If we could make this a const, then I'd be on board but unfortunately common.Address is not a const-compatible type.

File: messages/off-chain-registry/message_manager.go
Comment (cam-schultz) [reply messages/off-chain-registry/message_manager.go:85]: Since the tx bytes will be different, we'll actually successfully send the tx each time. It'd then be up to the TeleporterRegistry implementation to revert. That very well may be the case, but the relayer shouldn't depend on contract implementation details to function correctly.

We should definitely be checking if the address/version has already been registered, and abort relaying if so. I'll implement `ShouldSendMessage` to check for this. Thanks for pointing this out.

File: messages/off-chain-registry/message_manager.go
Comment (cam-schultz) [reply messages/off-chain-registry/message_manager.go:75]: Added

File: messages/off-chain-registry/message_manager.go
Comment (cam-schultz) [reply messages/off-chain-registry/message_manager.go]: Good call, removed.

File: go.mod
Comment (geoff-vball) [reply go.mod:9]: Okay thank you!

**Comment by michaelkaplan13** [comment]
Synced briefly offline, but commenting here for visibility that I'll defer reviewing this PR until we have further design discussions regarding whether off-chain warp message support in the relayer should be made contract-agnostic by having the configuration specify the raw input data of the transaction to be sent to deliver the message. 

This would allow for specifying the destination contract, function, and parameter values in the configuration, and would remove the need for any contract specific code in the relayer. The downside @cam-schultz called out is that the relayer then would not know how to check whether or not the message has been properly delivered or not, creating a risk that it could try to deliver the message multiple times if stuck in a restart loop with a off-chain message specified in its configuraiton.

**Comment by cam-schultz** [comment]
> Synced briefly offline, but commenting here for visibility that I'll defer reviewing this PR until we have further design discussions regarding whether off-chain warp message support in the relayer should be made contract-agnostic by having the configuration specify the raw input data of the transaction to be sent to deliver the message.
> 
> This would allow for specifying the destination contract, function, and parameter values in the configuration, and would remove the need for any contract specific code in the relayer. The downside @cam-schultz called out is that the relayer then would not know how to check whether or not the message has been properly delivered or not, creating a risk that it could try to deliver the message multiple times if stuck in a restart loop with a off-chain message specified in its configuration.

We came to the decision that this should be implemented as a separate `MessageManager` implementation. The main downside is that supporting generic messages limits how fine-grained `ShouldSendMessage` is able to be. Thus, generic message support and more "strongly typed" message support represent different use cases. I created https://github.com/ava-labs/awm-relayer/issues/165 to track the generic off-chain message use case

**Comment by geoff-vball** [comment]
Can you fix the merge conflicts here? Then I'll take a look

**Comment by cam-schultz** [comment]
> Can you fix the merge conflicts here? Then I'll take a look

Should be all set now

File: messages/off-chain-registry/message_manager.go
Comment (geoff-vball) [inline messages/off-chain-registry/message_manager.go:142]: Should this be an error log?

File: messages/off-chain-registry/message_manager.go
Comment (geoff-vball) [inline messages/off-chain-registry/message_manager.go:112]: Would it make sense for all the errors in this function to wrap the error we return with the error message, and then log the error from the calling function?

Comment (cam-schultz) [reply messages/off-chain-registry/message_manager.go:112]: We don't really have a formal logging strategy in this repo, but in general we err on the side of over rather than under logging. What that translates to is that most of the data structures are composed of `Logger`s, and log internally before returning an error, which may lead to the same error being logged again.

As for wrapping, I think that would be a good idea, but I also think it would bloat the source code significantly when combined with the above logging strategy. We'd have to log the error wrapping text and then also wrap the error itself with the same text before returning.

Thinking about how to unify this, we could implement our own `Logger`, adding a new method `WrapError` that 1) logs the wrapped error and 2) returns the wrapped error so that it can easily be propagated. Something like:

```
func (l *LoggerWrapper) WrapError(wrapText string, err error) error {
    wrappedErr := fmt.Errorf("%s: %w", wrapText, err)
    l.Error(
        wrapText,
        zap.Error(err),
    )
    return wrappedErr
}
```

Regardless, I think the current implementation in this PR is generally consistent with how we're handling error logging in the rest of the repo, so I'd prefer to defer to a separate PR in which we unify our error logging across the entire repo at once.

File: messages/off-chain-registry/message_manager.go
Comment (cam-schultz) [reply messages/off-chain-registry/message_manager.go]: Yep! Fixed.

File: messages/off-chain-registry/message_manager.go
Comment (geoff-vball) [reply messages/off-chain-registry/message_manager.go:112]: Yeah no problem for this PR. I've just been thinking about this a bit and would love a way to get the code size down a bit if possible.

**Review by geoff-vball — DISMISSED** [review]

File: messages/off-chain-registry/message_manager.go
Comment (michaelkaplan13) [inline messages/off-chain-registry/message_manager.go:126]: If there is any address registered for this versions already, should we not send the message?

File: messages/off-chain-registry/message_manager.go
Comment (michaelkaplan13) [inline messages/off-chain-registry/message_manager.go:69]: Could you add some unit tests for the new message manager implementation (particularly `ShouldSendMessage`)

File: messages/off-chain-registry/message_manager.go
Comment (cam-schultz) [reply messages/off-chain-registry/message_manager.go]: Yup, that's indeed the case. We only send the message if the version is not found. In that case, we don't care if the address is already registered, since the same address can map to multiple versions. We can therefor eliminate the call to GetVersionFromAddress altogether.

File: messages/off-chain-registry/message_manager.go
Comment (cam-schultz) [reply messages/off-chain-registry/message_manager.go:69]: Sure thing, added unit tests in for `ShouldSendMessage`. I omitted them for `SendMessage`, since that's basically a pass through to `DestinationClient.SendTx`, which already has unit tests.

File: messages/off-chain-registry/message_manager.go
Comment (cam-schultz) [reply messages/off-chain-registry/message_manager.go]: Yup, that's indeed the case. We only send the message if the version is not found. In that case, we don't care if the address is already registered, since the same address can map to multiple versions. We can therefor eliminate the call to GetVersionFromAddress altogether.

**Review by geoff-vball — APPROVED** [review]

**Review by michaelkaplan13 — APPROVED** [review]
LGTM!

Lets loop back to the longer term conversation of where it makes most sense to relay off-chain messages from soon-ish.

**Pull request merged**
